fuse/ipns/ipns_unix.go

```diff
@@ -336,21 +338,46 @@ func (s *Node) ReadDir(intr fs.Intr) ([]fuse.Dirent, fuse.Error) {
 	return nil, fuse.ENOENT
 }
 
-// ReadAll reads the object data as file data
-func (s *Node) ReadAll(intr fs.Intr) ([]byte, fuse.Error) {
-	log.Debugf("ipns: ReadAll [%s]", s.name)
-	r, err := uio.NewDagReader(context.TODO(), s.Nd, s.Ipfs.DAG)
+func (s *Node) Read(req *fuse.ReadRequest, resp *fuse.ReadResponse, intr fs.Intr) fuse.Error {
+	// intr will be closed by fuse if the request is cancelled. turn this into a context.
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel() // make sure all operations we started close.
+
+	// we wait on intr and cancel our context if it closes.
+	go func() {
+		select {
+		case <-intr: // closed by fuse
+			cancel() // cancel our context
+		case <-ctx.Done():
+		}
+	}()
+
+	k, err := s.Nd.Key()
 	if err != nil {
-		return nil, err
+		return err
 	}
-	// this is a terrible function... 'ReadAll'?
-	// what if i have a 6TB file? GG RAM.
-	b, err := ioutil.ReadAll(r)
+
+	// setup our logging event
+	lm := make(lgbl.DeferredMap)
+	lm["fs"] = "ipns"
+	lm["key"] = func() interface{} { return k.Pretty() }
+	lm["req_offset"] = req.Offset
+	lm["req_size"] = req.Size
+	defer log.EventBegin(ctx, "fuseRead", lm).Done()
+
+	r, err := uio.NewDagReader(ctx, s.Nd, s.Ipfs.DAG)
```

Comment:
I dont like that we have to do this... after the alpha im going to do a heavy refactor on the entire ipns subsystem and make it a lot better.